api/api/serializers/image_serializers.py

```diff
@@ -59,13 +60,31 @@ class ImageSearchRequestSerializer(
         required=False,
     )
 
-    def validate_internal__index(self, value):
-        index = super().validate_internal__index(value)
-        if index is None:
-            return None
-        if not index.startswith(IMAGE_TYPE):
-            raise serializers.ValidationError(f"Invalid index name `{value}`.")
-        return index
+
+class ImageSearchRequestSerializer(
+    ImageListRequestSerializer,
+    ImageSearchRequestSourceSerializer,
+    MediaSearchRequestSerializer,
```

Comment:
I know this was discussed, and I did go and follow the inheritance chain here to confirm that `MediaSearchRequestSerializer` actually isn't in the inheritance chain, but I can't help but feel that there's something a bit wonky going on with the composition of these types. I did a bit of thinking and came up with a solution I think is a good potential alternative, which is shared below the horizontal rule at the end of this comment. I also wrote two other long suggestions that I think are _bad_ and don't help matters and are just confusing in different ways. I didn't want to delete them becuase I think they're _tempting_ alternatives, and I wanted to share my thought process to anticipate different rabbit holes we could work ourselves into and might be able to avoid :slightly_smiling_face: .

<details>
<summary>Two bad alternatives</summary>
One thing that could help because it would (a) untangle the path params from the query params (which Staci pointed out as an issue) and (b) reduce the number of layers of inheritance, is to not use the same serializer for the path and the query parameters. For example, rather than having the collections serializer inherit from the list serializer (with all the query params), rename the "list" serializer to e.g., "MediaSearchFiltersSerializer", and keep the path parameters in a separate, non-heirarchical `MediaCollectionPathParamsSerializer`. The bonus is that this new approach would require no additional inheritance for the collection, and the `SearchQuery` serializer (renamed from "MediaSearchRequestSerializer") would just need to add those collection params.

Although, now that I'm writing it out, I think there's an even better approach: `FilterParamsSerializer`, `CollectionParamsSerializer`, which are individually used as the query and path param serializers respectively for the new collections endpoints and then combined into a `SearchParamsSerializer` used for the query params of the existing search endpoint.

This eliminates one class from these inheritance lists at the media type level. We'd have the following:

```py
# media_serializers.py
class FilterParamsSerializer:
    """Parameters accepted by all media search endpoints."""
    ...

class CollectionParamsSerializer:
    """
    Parameters used to narrow media by collection.

    Collections are defined as sources, creators, and tags.
    """
    ...

class SearchParamsSerializer(CollectionParamsSerializer, FilterParamsSerializer):
    """Base class for media type search parameters."""
    ...

# image_serializers.py
class ImageFilterParamsSerializer(FilterParamsSerializer):
    """Image-specific parameters for filtering images."""
    ...

class ImageSearchRequestSerializer(ImageFilterParamsSerializer, SearchParamsSerializer):
    """All-image search parameters"""
    ...

# audio_serializers.py
"""
Same as image
"""
```

Endpoint serializer configuration would be as follows:

* `/{media_type}/`: `*SearchRequestSerializer` only
* `/{media_type}/{collection_type}/{collection}/`: `*FilterParamsSerializer` for query params and `*CollectionParamsSerializer` for the path (actually doesn't make much sense :confused:, `collection_type` and `collection` are the keys here, not `creator`, `source`, and `tag`).

</details>

---

**Here's what I think is a _good_ alterative.**

After writing all that, I actually disagree with myself and have yet another suggestion: don't use the serializer for the collection path parameters at all. We'd still need to split the collection params out from the filter params, but we wouldn't need a `SearchRequest` base class as it does nothing to help reduce the code at the media type side. Rather than passing the collection filters as part of the serializer payload (artificially putting them into the `data`, pulling them from the path params), they would be passed as a dataclass with two keys: `collection_type` and `collection`.

In that case, we'd end up with:

* `FilterParamsSerializer`, as above
* `CollectionParamsSerializer`, as above
* No `MediaSearchRequestSerializer` base class anymore
* `{MediaType}FilterParamsSerializer` inherits from `FilterParamsSerializer` and adds media-type specific filters.
* `{MediaType}SearchParamsSerializer` inherits from `CollectionParamsSerializer` and `{MediaType}FilterParamsSerializer`.

For the routes:

* `/{media_type}/` uses `{MediaType}SearchParamsSerializer` for the query parameters
* `/{media_type}/{collection_type}/{collection}/` uses `{MediaType}FilterParamsSerializer` for the query params and passes a dataclass to search entrypoint to configure the collection aspect of the query.

I think this also requires slightly changing the way the search strategy is organised. Rather than a single entrypoint that switches on a strategy parameter (and therefore needs to accept the same parameters otherwise for both strategies, or else have very confusing parameters), have two separate search functions that share certain implementation details, for example:

```py
def search("""The current parameters"""):
    ...

def search_collection(Params<Search> + [collection_type: CollectionType, collection: str]):
    ...
```

Those two functions would then compose whatever shared behaviour was extracted from the current `search` function, rather than `search` needing to compose for both strategies.

I think the issue right now is that we're trying to compose from the inside out, rather than bottom up. This causes us to end up with really confusing inheritance trees like this one on the current `ImageSearchRequestSerializer`.